### ISSUE TITLE
Refactoring leads to incorrect code

### ISSUE THREAD

**Issue by karthiknadig** [pappasam/jedi-language-server#115]
This was originally filed on `vscode-python`, I closed the issue there and it should be addressed in this repo.

https://github.com/microsoft/vscode-python/issues/15802#issue-843437770

**Comment by pappasam** [comment]
tl;dr refactoring does exactly what it should in this case. That said, we might re-consider the way we recommend certain refactorings to users in vscode.

I tried this out on my machine, and while the suggestion appears to be nonsensical in the context of [scamp](https://github.com/MarcTheSpark/scamp), it's actually a surprisingly correct (and sophisticated) recommendation for inlining a variable.

Here, Jedi's `inline_variable` action is simply saying that, if the user wants, they can inline the variable `brass` throughout this module. In certain contexts, this might be desired behavior, but it appears that in this case, `s.new_part` creates a side effect within Session `s` and inlining the code to define the variable `brass` results in repeating this side-effect each time the inlined code is called.

Based on my analysis, this is less of an issue with jedi-language-server's handling of the `inline_variable` codeAction and more an issue with the vscode client's tone in recommending such an action. It turns out that it's **not** always a good idea to inline a variable. That said, the decision about whether or not to inline code is up to the software developer themselves (maybe, in some weird universe, a developer WANTS to repeat the side effect produced by `s.new_part`...).

Some possibilities I can think of:

- We chance the way suggested code actions are presented to users somehow, making it clearer what they do.
- We let users disable certain types of code actions

Am curious to hear your thoughts.

## Original

```python
import random

from scamp import *

s = Session()

brass = s.new_part("brass")
trumpet = s.new_part("trumpet")


def beat_it(transposition, short_length, medium_length, long_length):
    brass.play_note(40 + transposition, 1.0, short_length)
    brass.play_note(43 + transposition, 1.0, short_length)
    brass.play_note(47 + transposition, 1.0, short_length)
    brass.play_note(55 + transposition, 1.0, short_length)
    brass.play_note(52 + transposition, 1.0, long_length)
    brass.play_note(54 + transposition, 1.0, medium_length)
    brass.play_note(52 + transposition, 1.0, short_length)
    brass.play_note(50 + transposition, 1.0, short_length)
    wait(short_length)
    brass.play_note(50 + transposition, 1.0, short_length)
    wait(long_length)
```

## Code Action Result

```python
import random

from scamp import *

s = Session()

trumpet = s.new_part("trumpet")


def beat_it(transposition, short_length, medium_length, long_length):
    (s.new_part("brass")).play_note(40 + transposition, 1.0, short_length)
    (s.new_part("brass")).play_note(43 + transposition, 1.0, short_length)
    (s.new_part("brass")).play_note(47 + transposition, 1.0, short_length)
    (s.new_part("brass")).play_note(55 + transposition, 1.0, short_length)
    (s.new_part("brass")).play_note(52 + transposition, 1.0, long_length)
    (s.new_part("brass")).play_note(54 + transposition, 1.0, medium_length)
    (s.new_part("brass")).play_note(52 + transposition, 1.0, short_length)
    (s.new_part("brass")).play_note(50 + transposition, 1.0, short_length)
    wait(short_length)
    (s.new_part("brass")).play_note(50 + transposition, 1.0, short_length)
    wait(long_length)
```

Notice that the above code is technically the brass variable, inlined. This is a correct example of inlining a variable, it's just logically incorrect for the developer's intended behavior.

**Comment by karthiknadig** [comment]
@frmaurer99

**Comment by pappasam** [comment]
@karthiknadig would you consider this resolved? Not sure what to do specifically about this example given my explanation above.

**Comment by karthiknadig** [comment]
@pappasam  Thanks for the explanation. For now, I think we can close this.

@frmaurer99 Please let us know if there are any further questions or concerns.
